pkg/search/search.go

```diff
@@ -0,0 +1,214 @@
+	Devel        bool
+	Version      string
+	MaxColWidth  uint
+	RepoFile     string
+	RepoCacheDir string
+	OutputFormat output.Format
+}
+
+func (o *RepoOptions) Run(logger log.Logger, args []string) error {
+	wInfo := logio.NewWriter(logger, log.InfoLevel)
+	o.setupSearchedVersion()
+
+	index, err := o.buildIndex()
+	if err != nil {
+		return err
+	}
+
+	var res []*search.Result
+	if len(args) == 0 {
+		res = index.All()
+	} else {
+		q := strings.Join(args, " ")
+		res, err = index.Search(q, searchMaxScore, o.Regexp)
+		if err != nil {
+			return err
+		}
+	}
+
+	search.SortScore(res)
+	data, err := o.applyConstraint(res)
+	if err != nil {
+		return err
+	}
+
+	return o.OutputFormat.Write(wInfo, &repoSearchWriter{data, o.MaxColWidth})
+}
+
+func (o *RepoOptions) setupSearchedVersion() {
+	log.Debug("Original chart version: %q", o.Version)
+
+	if o.Version != "" {
+		return
+	}
+
+	if o.Devel { // search for releases and prereleases (alpha, beta, and release candidate releases).
+		log.Debug("setting version to >0.0.0-0")
+		o.Version = ">0.0.0-0"
+	} else { // search only for stable releases, prerelease versions will be skip
+		log.Debug("setting version to >0.0.0")
+		o.Version = ">0.0.0"
+	}
+}
+
+func (o *RepoOptions) applyConstraint(res []*search.Result) ([]*search.Result, error) {
+	if o.Version == "" {
+		return res, nil
+	}
+
+	constraint, err := semver.NewConstraint(o.Version)
+	if err != nil {
+		return res, errors.Wrap(err, "an invalid version/constraint format")
+	}
+
+	data := res[:0]
+	foundNames := map[string]bool{}
+	for _, r := range res {
+		// if not returning all versions and already have found a result,
+		// you're done!
+		if !o.Versions && foundNames[r.Name] {
+			continue
+		}
+		v, err := semver.NewVersion(r.Chart.Version)
+		if err != nil {
+			continue
+		}
+		if constraint.Check(v) {
+			data = append(data, r)
+			foundNames[r.Name] = true
+		}
+	}
+
+	return data, nil
+}
+
+func (o *RepoOptions) buildIndex() (*search.Index, error) {
+	// Load the repositories.yaml
+	rf, err := repo.LoadFile(o.RepoFile)
+	if isNotExist(err) || len(rf.Repositories) == 0 {
+		return nil, errors.New("no repositories configured")
+	}
+
+	i := search.NewIndex()
+	for _, re := range rf.Repositories {
+		n := re.Name
+		f := filepath.Join(o.RepoCacheDir, helmpath.CacheIndexFile(n))
+		ind, err := helmRepo.LoadIndexFile(f) // Needs to be helm type as we call the addrepo from helm directly
+		if err != nil {
+			log.Warn("Repo %q is corrupt or missing. Try 'helm repo update'.", n)
+			log.Warn("%s", err)
+			continue
+		}
+
+		i.AddRepo(n, ind, o.Versions || len(o.Version) > 0)
+	}
+	return i, nil
+}
+
+type repoChartElement struct {
+	Name        string `json:"name"`
+	Version     string `json:"version"`
+	AppVersion  string `json:"app_version"`
+	Description string `json:"description"`
+}
+
+type repoSearchWriter struct {
+	results     []*search.Result
+	columnWidth uint
+}
+
+func (r *repoSearchWriter) WriteTable(out io.Writer) error {
```

Comment:
Is it function used somewhere? (also, I don't see tests for it)

Reply:
This is part of  the `helm/pkg/cli/output/output.go` and called by line 91 in `search/search.go` based on the `RepoOptions.OutputFormat` which is set via command line flag as part of the `bindOutputFlag` :)

A bit convoluted, I agree but this is basically copied from helm.

Reply:
Thanks for the explanation!